applications/leetcoin.md

```diff
@@ -4,55 +4,118 @@
-3. Online judge engine - the core engine component of the backend app that executes the coding solutions in an isolated environment. The engine supports Ink!, Solidity and Vyper smart contract languages. It compiles the smart contract code, runs it and saves the output to a text file to a dedicated storage. The engine also reads the submission outputs from the storage and compares it against the expected output and reports back the failed and succeeded test cases along with the execution metadata (e.g statistics) and reports them back to the backend application.
+
+1. Frontend application - the UI React application that displays all the available problems, problem pages, code editor,
+   user profile, login/signup page, and admin dashboard. The frontend application interacts with the backend application
+   using REST API to request available coding problems, submit problem submissions, retrieve submission results,
+   retrieve user information, etc.
+2. Backend application - the core NodeJS application that is responsible for the business logic. The application has a
+   database that stores the list of available coding problems, users, coding submissions and etc. The main purpose is to
+   serve the data from the database via REST API.
+3. Online judge engine - the core engine component of the backend app that executes the coding solutions in an isolated
+   environment. The engine supports Ink!, Solidity and Vyper smart contract languages. It compiles the smart contract
+   code, runs it and saves the output to a dedicated storage in string format. The engine also reads the submission
+   outputs from the storage and compares it against the expected output and reports back the failed and succeeded test
+   cases along with the execution metadata (e.g statistics) and reports them back to the backend application.
    ![](https://imgur.com/fDx7QFR.png)
 
 #### Technology Stack
-The frontend application is written in React and hosted on AWS Amplify platform. The backend application utilizes express server framework which is built on top of NodeJS. It is really easy to design REST API using express and the API specification will be written in Swagger. TypeScript is used on the backend. The online judge is packaged as a portable module that can be easily incorporated in the other projects. The judge is written in TypeScript. The submission output database will be initially hosted on S3 as a bucket.
+
+The frontend application is written in React and NextJS and hosted on AWS Amplify platform. The backend application utilizes
+express server framework which is built on top of NodeJS. It is really easy to design REST API using express and the API
+specification will be written in Swagger. TypeScript is used on the backend. The online judge is packaged as a portable
+module that can be easily incorporated in the other projects. The judge is written in TypeScript. The submission output
+database will be initially hosted on Postgres.
```

Comment:
Previously, we considered S3 to host submission output but realized that Postgres satisfies all our requirements and makes it easier to develop, test locally and maintain in general. The output is stored as JSONB column